File: ocaml/bin/p0046.ml
(* Project Euler: Problem 46 *)

(*
  odd composite = prime? + 2 * n^2
 *)

open Core

let is_twice_square num =
  let tmp = Euler.Math.isqrt (num / 2) in
  if tmp * tmp = num / 2 then true else false

let rec find_counter_example n primes =
  if Euler.Math.is_prime n then
    find_counter_example (n + 2) (n :: primes)
  else
    let rec aux = function
      | hd :: _ when Bool.(is_twice_square (n - hd) = true) -> find_counter_example (n + 2) primes
      | _ :: tl -> aux tl
      | [] -> n
    in
    aux primes

let exec () =
  (* '9' is the smallest odd composite number. '[7;5;3]' are odd prime numbers less than '9'. *)
  Int.to_string (find_counter_example 9 [7; 5; 3])

let () = Euler.Task.run exec

File: rust/src/bin/p0046.rs
// Project Euler: Problem 46

euler::run_solver!(46);

fn solve() -> String {
    compute().to_string()
}

fn compute() -> i64 {
    use euler::math::{self, primes};

    fn is_twice_square(n: i64) -> bool {
        n % 2 == 0 && math::isqrt(n / 2).pow(2) == n / 2
    }

    // Two is not odd
    let mut odd_primes: Vec<i64> = vec![3, 5, 7, 11, 13, 17, 19, 23, 29, 31];

    for x in (35..).step_by(2) {
        if primes::is_prime(x) {
            odd_primes.push(x);
        } else if !odd_primes.iter().any(|&p| is_twice_square(x - p)) {
            return x;
        }
    }

    // Not reached on this problem
    unreachable!();
}

#[cfg(test)]
mod tests {
    use super::compute;

    #[test]
    fn p0046() {
        assert_eq!(compute(), 5777);
    }
}
